lib/instrumentation/http-shared.js

```diff
@@ -152,21 +153,15 @@ exports.traceOutgoingRequest = function (agent, moduleName, method) {
       // If so, then fallback to the current run context's span or transaction,
       // if any.
       var parent = span || parentRunContext.currSpan() || parentRunContext.currTransaction()
-      if (parent && parent._context) {
-        const headerValue = parent._context.toTraceParentString()
-        const traceStateValue = parent._context.toTraceStateString()
-
-        const newHeaders = Object.assign({}, options.headers)
-        newHeaders.traceparent = headerValue
-        if (traceStateValue) {
-          // Do not set the header to the empty string. Though RFC 7230 allows
-          // empty header values, some servers apparently do not.
-          newHeaders.tracestate = traceStateValue
-        }
-        if (agent._conf.useElasticTraceparentHeader) {
-          newHeaders['elastic-apm-traceparent'] = headerValue
-        }
-        options.headers = newHeaders
+      if (parent && parent._context && span) {
+        options.headers = span.propagateTraceContextHeaders(options.headers, parent._context)
+      } else if (parent && parent._context && !span) {
+        // headers still need to propagate
+        options.headers = propagateTraceContextHeaders(options.headers, parent._context)
+      }
```

Comment:
This is the kind of API I'd suggest:
```suggestion
      if (parent) {
         // uses the the TraceContext from parent._context
        parent.propagateTraceContextHeaders(
          // this argument (may be named `carrier`) is passed down to the header manipulating function
          req,
          // allows to pass in a technology specific function that handles the act of adding headers to a "header carrier" object
          // In the case of http, this will be the http request object
          // For other technologies, like SQS, it will be something more specific
          // This has the benefit that we don't need to allocate an intermediate map that holds the headers,
          // instead, we can directly set the headers on the request.
          // Note that we don't have to know about the names of the headers being added. Only the `TraceContext` class knows about the header names.
          // The thing that `TraceContext` doesn't have to know about is how to add a header for a specific technology.
          // Because the function receives the request object as a parameter, it allows us to store this function as a constant, avoiding having to allocate an anonymous function every time
          (req, key, value) => req.headers[key] = value
        )
      }
```

Reply:
That makes sense. Just to say it back, the semantics for "propagating" should change from "add the headers from this random context to an object" to "add _my_ headers to this object".  Also, rather than directly setting header values the new method will accept a callback so that the caller can define how headers are set.

Re: that callback -- I'll definitely build it into the API _but_ we'll still be allocating a new object for request headers.  We've [run into issues](https://github.com/elastic/apm-agent-nodejs/commit/5170e0f9ec4c40a02772a9c7d4fe291692c6516f) where modifying the headers object in place causes issues with AWS v4 signatures.

Reply:
> I'll definitely build it into the API but we'll still be allocating a new object for request headers. 

As APIs are harder to change and can limit the efficiency of all implementations, I think getting the API right and efficient is definitely more important than optimizing all implementations right now.